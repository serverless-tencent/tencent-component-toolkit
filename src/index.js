const Apigw = require('./modules/apigw');
const Cdn = require('./modules/cdn');
const Cns = require('./modules/cns');
const Cos = require('./modules/cos');
const Domain = require('./modules/domain');
const MultiApigw = require('./modules/multi-apigw');
const MultiScf = require('./modules/multi-scf');
const Scf = require('./modules/scf');
const Tag = require('./modules/tag');
const Postgresql = require('./modules/postgresql');
const Vpc = require('./modules/vpc');
const Cam = require('./modules/cam');
const Metrics = require('./modules/metrics');
const Layer = require('./modules/layer');

module.exports = {
  Apigw,
  Cdn,
  Cns,
  Cos,
  Domain,
  MultiApigw,
  MultiScf,
  Scf,
  Tag,
  Postgresql,
  Vpc,
  Cam,
  Metrics,
  Layer,
};
