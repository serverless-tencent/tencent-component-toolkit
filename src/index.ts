export { default as Apigw } from './modules/apigw';
export { default as Cdn } from './modules/cdn';
export { default as Cns } from './modules/cns';
export { default as Cos } from './modules/cos';
export { default as Domain } from './modules/domain';
export { default as Scf } from './modules/scf';
export { default as Tag } from './modules/tag';
export { default as Postgresql } from './modules/postgresql';
export { default as Vpc } from './modules/vpc';
export { default as Cam } from './modules/cam';
export { default as Metrics } from './modules/metrics/index';
export { default as Layer } from './modules/layer';
export { default as Cfs } from './modules/cfs';
export { default as Cynosdb } from './modules/cynosdb';
export { default as Cls } from './modules/cls';
export { default as Clb } from './modules/clb';
