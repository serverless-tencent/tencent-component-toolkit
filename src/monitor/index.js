const shimmer = require('./shimmer')
const Agent = require('./agent')

function initialize() {
  const agent = new Agent()
  // 封装 module的_load方法，在load时针对基础组件附加探针
  shimmer.patchModule()
  // 初始化一系列基础组件
  shimmer.bootstrapInstrumentation(agent)
}

initialize()
