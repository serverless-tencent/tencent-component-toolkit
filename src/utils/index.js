const path = require('path');
const fs = require('fs');

/**
 * simple deep clone object
 * @param {object} obj object
 */
function deepClone(obj) {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * return variable real type
 * @param {any} obj input variable
 */
function getRealType(obj) {
  return Object.prototype.toString.call(obj).slice(8, -1);
}

/**
 * is array
 * @param obj object
 */
function isArray(obj) {
  return Object.prototype.toString.call(obj) == '[object Array]';
}

/**
 * is object
 * @param obj object
 */
function isObject(obj) {
  return obj === Object(obj);
}

/**
 * iterate object or array
 * @param obj object or array
 * @param iterator iterator function
 */
function _forEach(obj, iterator) {
  if (isArray(obj)) {
    const arr = obj;
    if (arr.forEach) {
      arr.forEach(iterator);
      return;
    }
    for (let i = 0; i < arr.length; i += 1) {
      iterator(arr[i], i, arr);
    }
  } else {
    const oo = obj;
    for (const key in oo) {
      if (obj.hasOwnProperty(key)) {
        iterator(oo[key], key, obj);
      }
    }
  }
}

/**
 * flatter request parameter
 * @param source target object or array
 */
function flatten(source) {
  if (!isArray(source) && !isObject(source)) {
    return {};
  }
  const ret = {};
  const _dump = function(obj, prefix, parents) {
    const checkedParents = [];
    if (parents) {
      let i;
      for (i = 0; i < parents.length; i++) {
        if (parents[i] === obj) {
          throw new Error('object has circular references');
        }
        checkedParents.push(obj);
      }
    }
    checkedParents.push(obj);
    if (!isArray(obj) && !isObject(obj)) {
      if (!prefix) {
        throw obj + 'is not object or array';
      }
      ret[prefix] = obj;
      return {};
    }

    if (isArray(obj)) {
      // it's an array
      _forEach(obj, function(o, i) {
        _dump(o, prefix ? prefix + '.' + i : '' + i, checkedParents);
      });
    } else {
      // it's an object
      _forEach(obj, function(o, key) {
        _dump(o, prefix ? prefix + '.' + key : '' + key, checkedParents);
      });
    }
  };

  _dump(source, null);
  return ret;
}

function uniqueArray(arr) {
  return arr.filter((item, index, self) => {
    return self.indexOf(item) === index;
  });
}

function camelCase(str) {
  if (str.length <= 1) {
    return str.toUpperCase();
  }
  return `${str[0].toUpperCase()}${str.slice(1)}`;
}

function camelCaseProperty(obj) {
  let res = null;
  if (isObject(obj)) {
    res = {};
    Object.keys(obj).forEach((key) => {
      const val = obj[key];
      res[camelCase(key)] = isObject(val) || isArray(val) ? camelCaseProperty(val) : val;
    });
  }
  if (isArray(obj)) {
    res = [];
    obj.forEach((item) => {
      res.push(isObject(item) || isArray(item) ? camelCaseProperty(item) : item);
    });
  }
  return res;
}

function strip(num, precision = 12) {
  return +parseFloat(num.toPrecision(precision));
}

function traverseDirSync(dir, opts, ls) {
  if (!ls) {
    ls = [];
    dir = path.resolve(dir);
    opts = opts || {};
    if (opts.depthLimit > -1) {
      opts.rootDepth = dir.split(path.sep).length + 1;
    }
  }
  const paths = fs.readdirSync(dir).map((p) => dir + path.sep + p);
  for (let i = 0; i < paths.length; i++) {
    const pi = paths[i];
    const st = fs.statSync(pi);
    const item = { path: pi, stats: st };
    const isUnderDepthLimit =
      !opts.rootDepth || pi.split(path.sep).length - opts.rootDepth < opts.depthLimit;
    const filterResult = opts.filter ? opts.filter(item) : true;
    const isDir = st.isDirectory();
    const shouldAdd = filterResult && (isDir ? !opts.nodir : !opts.nofile);
    const shouldTraverse = isDir && isUnderDepthLimit && (opts.traverseAll || filterResult);
    if (shouldAdd) {
      ls.push(item);
    }
    if (shouldTraverse) {
      ls = traverseDirSync(pi, opts, ls);
    }
  }
  return ls;
}

module.exports = {
  isArray,
  isObject,
  _forEach,
  flatten,
  uniqueArray,
  camelCaseProperty,
  strip,
  traverseDirSync,
  getRealType,
  deepClone,
};
