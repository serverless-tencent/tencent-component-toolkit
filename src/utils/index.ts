import fs from 'fs';
import path from 'path';
import { PascalCase } from 'type-fest';

// TODO: 将一些库换成 lodash

/**
 * simple deep clone object
 * @param {object} obj object
 */
export function deepClone<T>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * return variable real type
 * @param {any} obj input variable
 */
export function getRealType<T>(obj: T): string {
  return Object.prototype.toString.call(obj).slice(8, -1);
}

/**
 * is array
 * @param obj object
 */
export function isArray<T>(obj: T[] | T): obj is T[] {
  return Object.prototype.toString.call(obj) == '[object Array]';
}

/**
 * is object
 * @param obj object
 */
export function isObject<T>(obj: T): obj is T {
  return obj === Object(obj);
}

/**
 * iterate object or array
 * @param obj object or array
 * @param iterator iterator function
 */
export function _forEach<T extends object>(
  obj: T[],
  iterator: (val: T, index: number, data?: T[]) => any,
): void;
export function _forEach<T extends object>(
  obj: T,
  iterator: (val: any, index: string, data?: T) => any,
): void;
export function _forEach<T extends object>(
  obj: T | T[],
  iterator: (val: any, index: any, data?: T | T[]) => any,
): void {
  if (isArray(obj)) {
    if (obj.forEach) {
      obj.forEach(iterator);
      return;
    }
    for (let i = 0; i < obj.length; i += 1) {
      iterator(obj[i], i, obj);
    }
  } else {
    for (const key in obj) {
      if (obj.hasOwnProperty(key)) {
        iterator(obj[key], key, obj);
      }
    }
  }
}

/**
 * flatter request parameter
 * @param source target object or array
 */
export function flatten<T extends object>(source: T | T[]) {
  if (!isArray(source) && !isObject(source)) {
    return {};
  }
  const ret: Record<string, any> = {};

  function _dump(obj: any, prefix?: string, parents?: any[]) {
    const checkedParents: any[] = [];
    if (parents) {
      let i;
      for (i = 0; i < parents.length; i++) {
        if (parents[i] === obj) {
          throw new Error('object has circular references');
        }
        checkedParents.push(obj);
      }
    }
    checkedParents.push(obj);
    if (!isArray(obj as any) && !isObject(obj)) {
      if (!prefix) {
        throw obj + 'is not object or array';
      }
      ret[prefix] = obj;
      return {};
    }

    if (isArray(obj)) {
      // it's an array
      _forEach(obj, function (o: any, i: number | string) {
        _dump(o, prefix ? prefix + '.' + i : '' + i, checkedParents);
      });
    } else {
      // it's an object
      _forEach(obj, function (o: any, key: string | number) {
        _dump(o, prefix ? prefix + '.' + key : '' + key, checkedParents);
      });
    }
  }

  _dump(source);
  return ret;
}

export function uniqueArray<T>(arr: T[]) {
  return arr.filter((item, index, self) => {
    return self.indexOf(item) === index;
  });
}

export function pascalCase<T extends string>(str: T): PascalCase<T> {
  if (str.length <= 1) {
    return str.toUpperCase() as any;
  }
  return `${str[0].toUpperCase()}${str.slice(1)}` as any;
}

export type PascalCasedProps<T> = {
  [K in keyof T as PascalCase<K>]: T[K] extends Array<infer U> | undefined
    ? Array<U>
    : PascalCasedProps<T[K]>;
};
export function pascalCaseProps<T>(obj: T): PascalCasedProps<T> {
  let res: Record<string, any> = {};
  if (isObject(obj)) {
    res = {} as any;
    Object.keys(obj).forEach((key: string) => {
      const val = (obj as any)[key];
      const k = pascalCase(key);
      res[k] = isObject(val) || isArray(val) ? pascalCaseProps(val) : val;
    });
  }
  if (isArray(obj as any)) {
    res = [];
    (obj as any).forEach((item: any) => {
      res.push(isObject(item) || isArray(item) ? pascalCaseProps(item) : item);
    });
  }
  return res as PascalCasedProps<T>;
}

export function strip(num: number, precision = 12) {
  return +parseFloat(num.toPrecision(precision));
}

export interface TraverseDirOptions {
  depthLimit?: number;
  rootDepth?: number;
  filter?: (item: { path: string; stats: fs.Stats }) => boolean;
  nodir?: boolean;
  nofile?: boolean;
  traverseAll?: boolean;
}

export function traverseDirSync(
  dir: string,
  opts?: TraverseDirOptions,
  ls?: { path: string; stats: fs.Stats }[],
): { path: string; stats: fs.Stats }[] {
  if (!ls) {
    ls = [];
    dir = path.resolve(dir);
    opts = opts ?? {};
    if (opts?.depthLimit ?? -1 > -1) {
      opts.rootDepth = dir.split(path.sep).length + 1;
    }
  }
  const paths: string[] = fs.readdirSync(dir).map((p: string) => dir + path.sep + p);
  for (let i = 0; i < paths.length; i++) {
    const pi = paths[i];
    const st = fs.statSync(pi);
    const item = { path: pi, stats: st };
    const isUnderDepthLimit =
      !opts?.rootDepth || pi.split(path.sep).length - opts.rootDepth < (opts?.depthLimit ?? -1);
    const filterResult = opts?.filter ? opts.filter(item) : true;
    const isDir = st.isDirectory();
    const shouldAdd = filterResult && (isDir ? !opts?.nodir : !opts?.nofile);
    const shouldTraverse = isDir && isUnderDepthLimit && (opts?.traverseAll || filterResult);
    if (shouldAdd) {
      ls?.push(item);
    }
    if (shouldTraverse) {
      ls = traverseDirSync(pi, opts, ls);
    }
  }
  return ls;
}
