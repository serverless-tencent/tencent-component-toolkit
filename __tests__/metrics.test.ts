import moment from 'moment';
import { Metrics } from '../src';

function format<T>(obj: T): void {
  if (Array.isArray(obj)) {
    (obj as Array<any>).sort();
    for (const v of obj) {
      format(v);
    }
  }

  if (typeof obj === 'object') {
    if (obj['type'] === 'timestamp') {
      obj['values'] = [];
    }
    for (const v of Object.values(obj)) {
      format(v);
    }
  }
}

describe('Metrics', () => {
  const credentials = {
    SecretId: process.env.TENCENT_SECRET_ID,
    SecretKey: process.env.TENCENT_SECRET_KEY,
  };
  const metrics = new Metrics(credentials, {
    funcName: 'serverless-unit-test',
  });

  const rangeStart = '2020-09-09 10:00:00';
  const rangeEnd = '2020-09-09 11:00:00';

  test('should get metrics data', async () => {
    const res = await metrics.getDatas(rangeStart, rangeEnd, 0xfffffffffff);
    expect(res).toEqual({
      rangeStart: moment(rangeStart).format('YYYY-MM-DD HH:mm:ss'),
      rangeEnd: moment(rangeEnd).format('YYYY-MM-DD HH:mm:ss'),
      metrics: expect.any(Array),
    });

    format(res);
    expect(res).toMatchSnapshot();
  });
});
