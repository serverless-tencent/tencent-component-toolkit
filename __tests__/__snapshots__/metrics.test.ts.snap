// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics should get metrics data 1`] = `
Object {
  "metrics": Array [
    Object {
      "title": "function invocations & errors",
      "type": "empty",
    },
    Object {
      "title": "function latency",
      "type": "empty",
    },
    Object {
      "title": "api requests & errors",
      "type": "stacked-bar",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "name": "requests",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
        Object {
          "color": "error",
          "name": "errors",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "api latency",
      "type": "multiline",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "name": "P95 latency",
          "total": 0,
          "type": "duration",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
        Object {
          "name": "P50 latency",
          "total": 0,
          "type": "duration",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "api 5xx errors",
      "type": "stacked-bar",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "color": "error",
          "name": "5xx",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "api 4xx errors",
      "type": "stacked-bar",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "color": "error",
          "name": "4xx",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "color": "error",
      "title": "api errors",
      "type": "list-flat-bar",
      "x": Object {
        "type": "string",
        "values": Array [
          "GET - /",
          "GET - /favicon.ico",
          "GET - /release",
          "POST - /",
          "POST - /upload",
          "POST - /upload/",
        ],
      },
      "y": Array [
        Object {
          "name": "404",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
        Object {
          "name": "500",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "api path requests",
      "type": "list-details-bar",
      "x": Object {
        "type": "string",
        "values": Array [
          "DELETE - /user/1",
          "GET - /",
          "GET - /css/app.0e433876.css",
          "GET - /css/vendor.74750554.css",
          "GET - /favicon.ico",
          "GET - /fonts/fluhrq6tzzclqej-vdg-iuiadsncihq8tq.b8c10426.woff2",
          "GET - /fonts/kfomcnqeu92fr1mu4mxm.49ae34d4.woff",
          "GET - /js/2.a13ff69c.js",
          "GET - /js/3.c5fd7573.js",
          "GET - /js/4.f6f9cf3e.js",
          "GET - /js/app.5a5abe2e.js",
          "GET - /js/vendor.b157e443.js",
          "GET - /playground",
          "GET - /playground/",
          "GET - /release",
          "GET - /user",
          "POST - /",
          "POST - /upload",
          "POST - /upload/",
          "POST - /user",
        ],
      },
      "y": Array [
        Object {
          "name": "requests",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
        Object {
          "name": "avg latency",
          "total": 0,
          "type": "duration",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "apigw total request num",
      "type": "stacked-bar",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "name": "request",
          "total": 0,
          "type": "count",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
    Object {
      "title": "apigw request response time(ms)",
      "type": "stacked-bar",
      "x": Object {
        "type": "timestamp",
        "values": Array [],
      },
      "y": Array [
        Object {
          "name": "response time",
          "total": 0,
          "type": "duration",
          "values": Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      ],
    },
  ],
  "rangeEnd": "2020-09-09 11:00:00",
  "rangeStart": "2020-09-09 10:00:00",
}
`;
